numba/targets/arraymath.py

```diff
@@ -517,6 +517,24 @@ def flat_any(a):
 
     return flat_any
 
+@overload(np.average)
+def np_average(arr, weights=None):
+    if weights is None or isinstance(weights, types.NoneType):
+        def np_average_impl(arr, weights=None):
+            return np.mean(arr)
+
+    else:
+        def np_average_impl(arr, weights=None):
```

Comment:
Don't forget that `arr` and `weights` could be `array_like`, it may be useful to call `np.asarray` on them as soon as possible to convert them to arrays so as to form a common type with which to work. Also, have a look at what the `dtype` of the result is for various input `dtypes`, for example, what happens if `weights` is a boolean array?

Reply:
@stuartarchibald regarding `dtype` is there a way/an example of setting the `dtype` in numba similar to https://github.com/numpy/numpy/blob/7ee52003/numpy/lib/function_base.py#L397-L400?

Reply:
You may find `numba.numpy_support.as_dtype` useful to call on arguments in the typing part of the implementation. This sort of pattern presented here is quite common:
https://github.com/numba/numba/blob/ccb57058d271bc0f8c151a48280a27dcd91c7f70/numba/targets/arraymath.py#L1509-L1514

Reply:
I've added a few new tests for cases where the input array and/or weights are boolean (https://github.com/numba/numba/pull/3655/commits/93eba0d7a7097cb4a02285df4bead5ac169b9080). Without casting with `numpy.promote_types` the output `dtype` is `float64`, so do we need to use type casting here?

Reply:
@slnguyen thanks for adding some test cases. In general the aim is to, where possible, try and match the type behaviour of NumPy. e.g. if NumPy returns a `int32` from a function taking a `float64` as an argument then Numba should do the same. Is there a specific example that is causing problems?